code/go/0chain.net/smartcontract/storagesc/models.go

```diff
@@ -1403,16 +1328,64 @@ func (sn *StorageAllocation) UnmarshalMsg(data []byte) ([]byte, error) {
 
 	*sn = StorageAllocation(*d)
 
-	sn.BlobberMap = make(map[string]*BlobberAllocation)
-	for _, blobberAllocation := range sn.BlobberDetails {
+	sn.BlobberAllocsMap = make(map[string]*BlobberAllocation)
+	for _, blobberAllocation := range sn.BlobberAllocs {
 		if blobberAllocation.Stats != nil {
 			sn.UsedSize += blobberAllocation.Stats.UsedSize // total used
 		}
-		sn.BlobberMap[blobberAllocation.BlobberID] = blobberAllocation
+		sn.BlobberAllocsMap[blobberAllocation.BlobberID] = blobberAllocation
 	}
 	return o, nil
 }
 
+// removeExpiredChallenges removes all expired challenges from the allocation,
+// return the expired challenge ids, or error if any.
+// the expired challenge ids could be used to delete the challenge node from MPT when needed
+func (sa *StorageAllocation) removeExpiredChallenges(allocChallenges *AllocationChallenges,
+	blobChallenges *BlobberChallenges, now common.Timestamp) ([]string, error) {
+	var (
+		expiredChallengeIDs           = make([]string, 0, len(allocChallenges.OpenChallenges))
+		lastExpiredBlobberChallengeID string
+	)
+
+	for _, oc := range allocChallenges.OpenChallenges {
+		ba, ok := sa.BlobberAllocsMap[oc.BlobberID]
+		if !ok {
+			return nil, common.NewErrorf("remove_expired_challenges", "blobber not exist in allocation: %s", oc.BlobberID)
+		}
+
+		// TODO: Not sure how the terms.ChallengeCompletionTime being set, perhaps we should get
+		// ChallengeCompletionTime from global config instead of the allocation's terms
+		expire := oc.CreatedAt + toSeconds(ba.Terms.ChallengeCompletionTime)
+		if expire > now {
+			// not expired, following open challenges would not expire too, so break here
+			break
+		}
+
+		// expired
+		expiredChallengeIDs = append(expiredChallengeIDs, oc.ID)
+
+		// update challenge stats
+		ba.Stats.FailedChallenges++
+		ba.Stats.OpenChallenges--
+
+		sa.Stats.FailedChallenges++
+		sa.Stats.OpenChallenges--
+
+		if oc.BlobberID == blobChallenges.BlobberID {
+			lastExpiredBlobberChallengeID = oc.ID
+		}
+	}
+
+	allocChallenges.OpenChallenges = allocChallenges.OpenChallenges[len(expiredChallengeIDs):]
+
+	if lastExpiredBlobberChallengeID != "" {
+		blobChallenges.removeChallengesTo(lastExpiredBlobberChallengeID)
```

Comment:
The idea was if the challenge is expired, then all challenges that prior to it must have expired too, but yes, it will cause consistent issue. I'll fix it.